modules/thread-pings/src/main/java/net/neoforged/camelot/module/threadpings/ThreadPingsListener.java

```diff
@@ -40,39 +40,43 @@ public void onEvent(@NotNull GenericEvent gevent) {
         if (!(event.isFromType(ChannelType.GUILD_PUBLIC_THREAD)))
             return;
-        
+
         final ThreadChannel thread = event.getChannel().asThreadChannel();
-        final List<Long> roleIds = new ArrayList<>();
-        database.useExtension(ThreadPingsDAO.class, threadPings -> {
-            // Check the thread's parent channel
-            final IThreadContainerUnion parentChannel = thread.getParentChannel();
-            roleIds.addAll(threadPings.query(parentChannel.getIdLong()));
-
-            if (parentChannel instanceof StandardGuildChannel guildChannel) {
-                // Check the category of the thread's parent channel
-                final Category parentCategory = guildChannel.getParentCategory();
-                if (parentCategory != null) {
-                    roleIds.addAll(threadPings.query(parentCategory.getIdLong()));
-                }
-            }
+        final List<Long> pingRoleIds = new ArrayList<>();
+        database.useExtension(ThreadPingsDAO.class,
+                threadPings -> queryRoles(thread, pingRoleIds, threadPings::query));
 
-            // Check guild-wide (using the guild ID)
-            roleIds.addAll(threadPings.query(thread.getGuild().getIdLong()));
-        });
+        final List<Long> exemptRoleIds = new ArrayList<>();
 
-        final List<Role> roles = new ArrayList<>();
-        for (Long roleId : roleIds) {
+        database.useExtension(ThreadPingsExemptionsDAO.class,
+                threadPingsExempt -> queryRoles(thread, exemptRoleIds, threadPingsExempt::query));
+
+        final List<Role> pingRoles = new ArrayList<>();
+        for (Long roleId : pingRoleIds) {
             final Role role = thread.getGuild().getRoleById(roleId);
             if (role == null) {
-                LOGGER.info("Role {} does not exist; deleting role from database", roleId);
+                LOGGER.info("Ping role {} does not exist; deleting role from database", roleId);
                 database.useExtension(ThreadPingsDAO.class, threadPings -> threadPings.clearRole(roleId));
                 continue;
             }
-            roles.add(role);
+            pingRoles.add(role);
+        }
+
+        final List<Role> exemptRoles = new ArrayList<>();
+        for (Long roleId : exemptRoleIds) {
+            final Role role = thread.getGuild().getRoleById(roleId);
+            if (role == null) {
+                LOGGER.info("Exempt role {} does not exist; deleting role from database", roleId);
+                database.useExtension(ThreadPingsExemptionsDAO.class, threadPingsExempt -> threadPingsExempt.clearRole(roleId));
+                continue;
+            }
+            exemptRoles.add(role);
         }
 
-        if (roles.isEmpty()) return;
-        final String mentionMessage = roles.stream()
+        pingRoles.removeAll(exemptRoles);
+
+        if (pingRoles.isEmpty()) return;
+        final String mentionMessage = pingRoles.stream()
```

Comment:
I think you'll want a distinct here or to turn pingRoleIds into a set given that it seems like it is possible for duplicate roles to end up in it if the same role is added at multiple levels.